lib/logstash/inputs/rabbitmq.rb

```diff
@@ -135,6 +137,60 @@ def shutdown_consumer
         @consumer.gracefully_shut_down
       end
 
+      private
+
+      def get_header_value(value)
+        # Two kinds of values require exceptional treatment:
+        #
+        # String values are instances of
+        # com.rabbitmq.client.impl.LongStringHelper.ByteArrayLongString
+        # and we don't want to propagate those.
+        #
+        # List values are java.util.ArrayList objects and we need to
+        # recurse into them to convert any nested strings values.
+        if value.respond_to? "toArray"
+          value.map{|item| get_header_value(item) }
+        elsif value.respond_to? "toString"
+          value.toString
+        else
+          value
+        end
+      end
+
+      private
+      def get_headers(metadata)
+        if !metadata.headers.nil?
+          Hash[metadata.headers.map {|k, v| [k, get_header_value(v)]}]
+        else
+          {}
+        end
+      end
+
+      private
+      def get_properties(metadata)
+        result = {}
+        # Technically the exchange, redeliver, and routing-key
+        # properties belong to the envelope and not the message but we
+        # ignore that distinction here. However, we extract the
+        # headers separately via get_headers even though the header
+        # table technically is a message property.
+        ["app-id", "cluster-id", "consumer-tag", "content-encoding",
+         "content-type", "correlation-id", "delivery-mode",
+         "exchange", "expiration", "message-id", "priority",
+         "redeliver", "reply-to", "routing-key", "timestamp",
+         "type", "user-id"].reduce({}) do |acc, name|
+
+          # The method names obviously can't contain hyphens.
+          value = metadata.send(name.gsub("-", "_"))
+          if value
+            # The AMQP 0.9.1 timestamp field only has second resolution
+            # so storing milliseconds serves no purpose and might give
+            # the incorrect impression of a higher resolution.
+            result[name] = name != "timestamp" ? value : value.getTime / 1000
```

Comment:
Feels odd to use a ternary for this special case. Maybe a separate call `result["timestamp"] = metadata.timestamp` outside of the loop instead?